docker build sometimes causes a panic (nil pointer)

This happens occasionally for me, in particular when building multiple images in parallel. It's not super deterministic, but our build process involves building lots of images relatively frequently, so we do it in parallel to be a bit faster.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x603c5f]

goroutine 1 [running]:
github.com/docker/docker/api/client.validateContextDirectory.func1(0xc8208e9c50, 0x26, 0x0, 0x0, 0x7f9eb2d0d880, 0xc8208e9dd0, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/client/build.go:295 +0x14f
path/filepath.walk(0xc82066dab0, 0x6, 0x7f9eb2d0d8a8, 0xc82066e780, 0xc820b3f278, 0x0, 0x0)
	/usr/local/go/src/path/filepath/path.go:370 +0x415
path/filepath.Walk(0xc82066dab0, 0x6, 0xc820b3f278, 0x0, 0x0)
	/usr/local/go/src/path/filepath/path.go:396 +0xe1
github.com/docker/docker/api/client.validateContextDirectory(0xc82066d630, 0x6, 0xc82008d800, 0x17, 0x20, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/client/build.go:325 +0xde
github.com/docker/docker/api/client.(*DockerCli).CmdBuild(0xc820240580, 0xc82000a0d0, 0x5, 0x5, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/client/build.go:152 +0x1fcf
reflect.callMethod(0xc82066ad50, 0xc820b3fc58)
	/usr/local/go/src/reflect/value.go:628 +0x1fc
reflect.methodValueCall(0xc82000a0d0, 0x5, 0x5, 0x1, 0xc82066ad50, 0x0, 0x0, 0xc82066ad50, 0x0, 0x473414, ...)
	/usr/local/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc82066aa20, 0xc82000a0c0, 0x6, 0x6, 0x0, 0x0)
	/go/src/github.com/docker/docker/cli/cli.go:89 +0x383
main.main()
	/go/src/github.com/docker/docker/docker/docker.go:63 +0x400

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 6 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
```
**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64
```


**Output of `docker info`:**

```
Containers: 10
 Running: 8
 Paused: 0
 Stopped: 2
Images: 304
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 308
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.2.0-35-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 2.872 GiB
Name: ubuntu-14
ID: HZUB:B5T5:TAFP:YCQJ:CLOB:ZQJG:7UE2:YVF6:RVXY:NRBQ:ERWI:G6X2
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Username: tonicjenkins
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running in a Parallels VM (Ubuntu 14.04)

**Steps to reproduce the issue:**

Non-deterministic; fire off a lot of parallel builds.

**Describe the results you received:**

Panic (trace above)

**Describe the results you expected:**

Build to complete successfully.

Issue still exists: reproduced this time in 1.11

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x81ab9f]

goroutine 1 [running]:
github.com/docker/docker/builder.ValidateContextDirectory.func1(0xc8201f9700, 0x75, 0x0, 0x0, 0x7f9eeb36b418, 0xc8202b0a80, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/builder/context.go:38 +0x14f
path/filepath.walk(0xc820339cc0, 0x31, 0x7f9eeb36b440, 0xc8203a12c0, 0xc82014d258, 0x0, 0x0)
	/usr/local/go/src/path/filepath/path.go:370 +0x415
path/filepath.Walk(0xc820339cc0, 0x31, 0xc82014d258, 0x0, 0x0)
	/usr/local/go/src/path/filepath/path.go:396 +0xe1
github.com/docker/docker/builder.ValidateContextDirectory(0xc820339a40, 0x31, 0xc82038ea00, 0x17, 0x20, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/builder/context.go:68 +0xde
github.com/docker/docker/api/client.(*DockerCli).CmdBuild(0xc8201f9c80, 0xc82000a0d0, 0x5, 0x5, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/api/client/build.go:144 +0x1e6c
reflect.callMethod(0xc820390a20, 0xc82014dc58)
	/usr/local/go/src/reflect/value.go:628 +0x1fc
reflect.methodValueCall(0xc82000a0d0, 0x5, 0x5, 0x1, 0xc820390a20, 0x0, 0x0, 0xc820390a20, 0x0, 0x47cf14, ...)
	/usr/local/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc8203906f0, 0xc82000a0c0, 0x6, 0x6, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/cli.go:89 +0x383
main.main()
	/usr/src/docker/docker/docker.go:63 +0x43c

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 6 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
```

```
root@ip-172-31-35-216:~# docker version
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64
```

--- a/builder/context.go
+++ b/builder/context.go
@@ -29,6 +29,16 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
 		return err
 	}
 	return filepath.Walk(contextRoot, func(filePath string, f os.FileInfo, err error) error {
+		if err != nil {
+			if os.IsPermission(err) {
+				return fmt.Errorf("can't stat '%s'", filePath)
+			}
+			if os.IsNotExist(err) {
+				return nil
+			}
+			return err
+		}
+
 		// skip this directory/file if it's not in the path, it won't get added to the context
 		if relFilePath, err := filepath.Rel(contextRoot, filePath); err != nil {
 			return err
@@ -39,16 +49,6 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
 				return filepath.SkipDir
 			}
 			return nil
-		}
-
-		if err != nil {
-			if os.IsPermission(err) {
-				return fmt.Errorf("can't stat '%s'", filePath)
-			}
-			if os.IsNotExist(err) {
-				return nil
-			}
-			return err
 		}
 
 		// skip checking if symlinks point to non-existing files, such symlinks can be useful

